src/pybind/mgr/mirroring/fs/snapshot_mirror.py

```diff
@@ -0,0 +1,579 @@
+                if r < 0:
+                    log.error(f'mon command to remove peer failed: {err}')
+                    raise Exception(-errno.EINVAL)
+                return 0, json.dumps({}), ''
+        except MirrorException as me:
+            return me.args[0], '', me.args[1]
+        except Exception as e:
+            return e.args[0], '', 'failed to remove peer'
+
+    def verify_and_normalize_dir(self, fs_handle, dir_path):
+        norm_path = os.path.normpath(dir_path)
+        try:
+            stx = fs_handle.statx(norm_path, cephfs.CEPH_STATX_MODE, cephfs.AT_SYMLINK_NOFOLLOW)
+            if not stat.S_ISDIR(stx["mode"]):
+                raise MirrorException(-errno.EINVAL, f'path {dir_path} should be a directory')
+        except cephfs.ObjectNotFound:
+            raise MirrorException(-errno.ENOENT, f'{dir_path} does not exist')
+        except cephfs.Error as e:
+            log.error(f'failed to verify directory {dir_path}: {e}')
+            raise Exception(-e.errno)
+        return norm_path
+
+    def verify_and_normalize_path(self, filesystem, dir_path):
+        if not os.path.isabs(dir_path):
+            raise MirrorException(-errno.EINVAL, f'{dir_path} should be an absolute path')
+        try:
+            with open_filesystem(self.local_fs, filesystem) as local_fs_handle:
+                return self.verify_and_normalize_dir(local_fs_handle, dir_path)
+        except CephfsConnectionException as e:
+            raise MirrorException(-e.errno, e.error_str)
+
+    def add_directory(self, filesystem, dir_path):
+        try:
+            with self.lock:
+                if not self.filesystem_exist(filesystem):
+                    raise MirrorException(-errno.ENOENT, f'filesystem {filesystem} does not exist')
+                fspolicy = self.pool_policy.get(filesystem, None)
+                if not fspolicy:
+                    raise MirrorException(-errno.EINVAL, f'filesystem {filesystem} is not mirrored')
+                dir_path = self.verify_and_normalize_path(filesystem, dir_path)
+                log.debug(f'path normalized to {dir_path}')
+                fspolicy.add_directory(dir_path)
+                return 0, json.dumps({}), ''
+        except MirrorException as me:
+            return me.args[0], '', me.args[1]
+        except Exception as e:
+            return e.args[0], '', 'failed to add directory'
+
+    def remove_directory(self, filesystem, dir_path):
+        try:
+            with self.lock:
+                if not self.filesystem_exist(filesystem):
+                    raise MirrorException(-errno.ENOENT, f'filesystem {filesystem} does not exist')
+                fspolicy = self.pool_policy.get(filesystem, None)
+                if not fspolicy:
+                    raise MirrorException(-errno.EINVAL, f'filesystem {filesystem} is not mirrored')
+                dir_path = self.verify_and_normalize_path(filesystem, dir_path)
+                fspolicy.remove_directory(dir_path)
```

Comment:
nit: I think, attempting to remove an un-added directory should not test the presence of the directory in the file-system
The list of added directories should be checked instead.

Also, what if the directory is added for mirroring and then deleted from file-system. In such a case, the user will not be able to remove the directory from the list of added directories.

Reply:
that's right -- but before we do this, the mirror daemon needs to be taught to handle missing directory failures. Right now, the mirror daemon would retry (after failure) and maybe just avoiding the check here would suffice. However, this needs to be tested. So, when I get to fixing the mirror daemon, this check would be removed. 

Good catch!